batch/sql/add-user-resources-ready-cores-token.sql

```diff
@@ -0,0 +1,176 @@
+SET @n_tokens = 200;
+
+ALTER TABLE globals ADD COLUMN n_tokens INT NOT NULL;
+UPDATE globals SET n_tokens = @n_tokens;
+
+CREATE TABLE IF NOT EXISTS `batches_staging` (
+  `batch_id` BIGINT NOT NULL,
+  `token` INT NOT NULL,
+  `n_jobs` INT NOT NULL DEFAULT 0,
+  `n_ready_jobs` INT NOT NULL DEFAULT 0,
+  `ready_cores_mcpu` BIGINT NOT NULL DEFAULT 0,
+  PRIMARY KEY (`batch_id`, `token`),
+  FOREIGN KEY (`batch_id`) REFERENCES batches(`id`) ON DELETE CASCADE
+) ENGINE = InnoDB;
+
+ALTER TABLE user_resources ADD COLUMN token INT NOT NULL DEFAULT 0;
+ALTER TABLE user_resources DROP PRIMARY KEY, ADD PRIMARY KEY(`user`, `token`);
+
+ALTER TABLE ready_cores ADD COLUMN token INT NOT NULL DEFAULT 0;
+ALTER TABLE ready_cores ADD PRIMARY KEY(`token`);
+
+CREATE TEMPORARY TABLE tmp_resources AS (
+  SELECT batch_id, user, closed,
+    0 as token,
+    COUNT(*) AS n_jobs,
+    SUM(state = 'Ready') AS n_ready_jobs,
+    SUM(state = 'Running') AS n_running_jobs,
+    SUM(IF(state = 'Ready', cores_mcpu, 0)) AS ready_cores_mcpu,
+    SUM(IF(state = 'Running', cores_mcpu, 0)) AS running_cores_mcpu
+  FROM jobs
+  INNER JOIN batches ON batches.id = jobs.batch_id
+  GROUP BY batch_id, user, closed
+);
+
+SELECT COALESCE(SUM(ready_cores_mcpu), 0) INTO @closed_ready_cores_mcpu
+FROM tmp_resources
+WHERE closed;
+
+UPDATE ready_cores SET ready_cores_mcpu = @closed_ready_cores_mcpu;
+
+UPDATE user_resources
+RIGHT JOIN (
+  SELECT user, token,
+    SUM(n_ready_jobs) as n_ready_jobs,
+    SUM(n_running_jobs) as n_running_jobs,
+    SUM(ready_cores_mcpu) as ready_cores_mcpu,
+    SUM(running_cores_mcpu) as running_cores_mcpu
+  FROM tmp_resources
+  WHERE closed
+  GROUP BY user, token) AS t
+ON user_resources.user = t.user AND user_resources.token = t.token
+SET
+  user_resources.n_ready_jobs = t.n_ready_jobs,
+  user_resources.n_running_jobs = t.n_running_jobs,
+  user_resources.ready_cores_mcpu = t.ready_cores_mcpu,
+  user_resources.running_cores_mcpu = t.running_cores_mcpu;
+
+INSERT INTO batches_staging (batch_id, token, n_jobs, n_ready_jobs, ready_cores_mcpu)
+  SELECT tmp_resources.batch_id,
+    tmp_resources.token,
+    tmp_resources.n_jobs,
+    tmp_resources.n_ready_jobs,
+    tmp_resources.ready_cores_mcpu
+  FROM tmp_resources
+  WHERE NOT closed;
+
+DELIMITER $$
+
+DROP TRIGGER IF EXISTS jobs_after_update;
+CREATE TRIGGER jobs_after_update AFTER UPDATE ON jobs
+FOR EACH ROW
+BEGIN
+  DECLARE in_user VARCHAR(100);
+  DECLARE rand_token INT;
+
+  SELECT user INTO in_user from batches
+  WHERE id = NEW.batch_id;
+
+  SET rand_token = FLOOR(RAND() * @n_tokens);
```

Comment:
You should query the globals to get this.  If we change the global, we want this to change, too.

Reply:
If I do that, then I need a lock on the globals table. Won't that cause the same problems we've had before where we're locking a single row in a table for everything? I'll change it though because you're correct it should be tied to the globals value.

Reply:
I can't actually find in the documentation what (if any) lock SELECT with no modifier has.  But it should either be no lock or a read lock.  Multiple transactions can hold a read lock simultaneously, so this shouldn't decrease concurrency.  If we're worried, we can be explicit with SELECT ... LOCK IN SHARE MODE (although no lock would be (slightly) preferable since we don't really care about synchronization on globals).

"Using LOCK IN SHARE MODE sets a shared lock that permits other transactions to read the examined rows but not to update or delete them."

The previous problem was taking a write lock, e.g. doing an UPDATE.

Reply:
From https://dev.mysql.com/doc/refman/5.7/en/innodb-locking-reads.html:

"If you query data and then insert or update related data within the same transaction, the regular SELECT statement does not give enough protection. Other transactions can update or delete the same rows you just queried."

I wish the documentation was slightly more explicit, but this really sounds like unmodified SELECT takes no locks.